### ISSUE TITLE
Some WASI imported functions need to write to memory

### ISSUE THREAD

**Issue by myobie** [tessi/wasmex#13]
An example of a WASI function that needs to be able to write to memory is [`random_get`](https://github.com/WebAssembly/WASI/blob/d5b0c23ed2730262d4eaac2b2721aad07c6f65b4/phases/snapshot/docs.md#-random_getbuf-pointeru8-buf_len-size---errno). This means one ends up in a position where one needs to pass a function as an import that can later gain access to the instance's memory. I've had to deal with this in other languages as well.

I wrote [an example test](https://gist.github.com/myobie/94ca667c1216b2fae3b72a7d72446220) and the error I see is:

```
15:23:54.272 [error] GenServer #PID<0.448.0> terminating
** (stop) exited in: GenServer.call(#PID<0.448.0>, {:memory, :uint8, 0}, 5000)
    ** (EXIT) process attempted to call itself 
```

I understand that this is a pretty strange expectation for WASI and [they are aware of it](https://github.com/WebAssembly/WASI/issues/185), but it doesn't appear a solution will arise soon IMO. It could be useful to pass the memory into imported functions as a kind of state. What do you think?

**Comment by tessi** [comment]
Thanks for opening this well prepared issue 💚 I agree in that we should tackle it.

We should pass memory (and maybe even more context) to callback functions. This is already done for `DynamicFunc`s for us in Rust-land ([we just chose to ignore the `vm::Ctx` at the current implementation](https://github.com/tessi/wasmex/blob/55f1c4b248fcbc12269800d91c79d0b318d835d4/native/wasmex/src/namespace.rs#L48)).

`vm::Ctx` offers us the [`memory`](https://docs.rs/wasmer-runtime/0.16.2/wasmer_runtime/struct.Ctx.html#method.memory) function which we could convert into a memory object and pass down to elixir.

We need to rework how memory resources function, though. Currently it remembers the instance and, for each task, asks it for its exports, finds the WASM-land memory, asks for the specified view (datatype and offset). Instead we need to save the memory object directly in the resource.

It should not be too hard to change the existing memory code. I would suggest to tackle this in three steps:

1. finish the import feature without passing memory in
2. (could be in parallel) rework the existing memory resource to hold a WASM memory instead of WASM instance
3. pass the memory into elixir callbacks for imported functions

About 3: I guess we have to change the method signature of elixir-callbacks here. Maybe they should receive a tuple of two elements:

* context - a map with `%{memory: Wasmex.Memory.t()}` and maybe later, WASM tables or other context-related infos)
* and the normal `params` array we pass in now

@myobie what do you think?

**Comment by tessi** [comment]
☝️ that being said, I _think_ you could get around the GenServer process attempting to call itself by not using the `Wasmex` GenServer interface, but implement your own - this way you would not need to call the GenServer just to retrieve the memory.

It is a workaround I don't like (because I want the higher-level GenServer interface to work), but it may get you a step further in your WASI experiement 🤞

**Comment by myobie** [comment]
Your notes and suggested steps make sense and I agree finishing the import feature as is is the top priority. 

I did start down the road of a custom GenServer this morning but realized it was too far.

**Comment by tessi** [comment]
I will add a TODO to #9 to change the signature of elixir callbacks to not only receive `params` but `{context, params}` (where context will be the empty map for now).

This is hoping that people have an easier time migrating to the bright future of us passing memory and other context into callbacks.

**Comment by tessi** [comment]
@myobie could you give the current `master` another try? I added support for memory access to elixir callbacks in #16

**Comment by myobie** [comment]
@tessi it works for my use  case for `random_get`. An example that works for me is:

```ex
{:ok, instance} =
  Wasmex.start_link(%{
    bytes: bytes,
    imports: %{
      wasi_snapshot_preview1: %{
        random_get:
          {:fn, [:i32, :i32], [:i32],
           fn %{memory: memory}, address, size ->
             bytes = :crypto.strong_rand_bytes(size)

             bytes
             |> :binary.bin_to_list()
             |> Enum.with_index()
             |> Enum.each(fn {byte, index} ->
               IO.puts("setting #{byte} at #{address + index}")
               Wasmex.Memory.set(memory, address + index, byte)
             end)

             0
           end}
      }
    }
  })
```

I can't use the function to write a binary because I don't want it to end with a null byte. I'm not sure  if it's worth having two functions: one for write binary and one for write null terminated string.

**Comment by tessi** [comment]
Good point with the string writing. After some reading and getting more proficient in Rust it became clear to me that we shouldn't rely on null-bytes in our string helpers. We should always use the memory `index` and the strings `length` to read or write memory (as it is done by Rust in the `String` type).

Sooner or later I'll change this API to not use null-bytes, but lengths. Feel free to (but no need to) contribute :)

**Comment by tessi** [comment]
In other news: You may have a look at #17 (and a WIP PR at #18) that adds WASI support. I plan to optionally pre-fill the imports object with WASI implementations written in Rust. They should ideally be overwritable from elixir if you want to implement some functions yourself.

Maybe this is helping you in your plans.

**Comment by myobie** [comment]
IMO this issue can be closed now that #16 has been merged.
